torch/distributed/distributed_c10d.py

```diff
@@ -1465,6 +1466,128 @@ def reduce_scatter(output,
         work.wait()
 
 
+def all_to_all(output,
```

Comment:
do you mind splitting this API into two: one for single tensor with output_split_sizes and one for list of tensors? 
This would be good if we want to do type safety checking in PY, e.g. mypy and other API in this file more or less follow this convention (inputs/outputs are either single tensors, list of tensors or list of lists of tensors)

Reply:
To be consistent with other APIs, I would prefer to keep `all_to_all` name for list variant. What name would you suggest for single tensor variant? `all_to_all_base` or `all_to_all_single` or something else?

Reply:
maybe all_to_all and all_to_all_single?

Reply:
Hi @mrshenli and @agolynski 

Is there an easy way to detect if a list of tensors is created from one big tensor using `split` or `chunk` operation? If yes, we can potentially eliminate need for `all_to_all_single`. I see that they share the same underlying storage but don't know if we can use any specific tensor property to infer that. The need for `all_to_all_single` comes from extra copies involved to create contiguous tensor inside MPI backend. If we can somehow avoid that, we can eliminate need for single tensor based version.

Reply:
We could do it, but I think it's good for consistency to have single tensor all-to-all API.

if you are interested in detecting such tensors you can take a look at https://github.com/pytorch/pytorch/pull/33924

Reply:
Thanks @agolynski . I had almost reinvented the same thing though I was trying to more generalize it to support uneven sizes and also different storage base, if possible, using offsets parameter available in most of vector versions of MPI APIs. (e.g. `MPI_Allgatherv`, `MPI_Alltoallv`). But I wasn't sure about right way to check. This PR confirms that.

Reply:
> maybe all_to_all and all_to_all_single?

Done!